e2e: Add SyncSource.Path

karlkfi opened the pull request:
- Add Path method that wraps Git & OCI dir & Helm chart name.
- Add SetExpectedSyncPath to update the directory expectation.
- Add SetExpectedGitSource, SetExpectedHelmSource, &
  SetExpectedOCISource to update source expectations and replace
  WithSyncDirectoryMap.
- Move all Git repo test defaults to setupTestCase
- Remove obsolete HelmChartVersionShaFn
- Add kinds.LookupID to replace usace of core.IDof using the schema.
- Add OCIImageID to simplify address/url generation
- Add RootSyncObjectGit and RepoSyncObjectGit to handle setting
  defaults that were in multiple places before.
- Add RootSyncObjectOCI & RepoSyncObjectOCI for building RSyncs with
  OCI sources.
- Add RootSyncObjectHelm & RepoSyncObjectHelm for building RSyncs
  with Helm sources.

nan-yu commented on e2e/testcases/workload_identity_test.go:
```diff
@@ -598,8 +583,27 @@ func updateRSyncWithGitSourceConfig(nt *nomostest.NT, rs client.Object, repo *gi
+}
+
+// TODO: promote to helper function, or find existing alternative
+func lookupObjectID(obj client.Object, scheme *runtime.Scheme) (core.ID, error) {
```
What about using `core.IDOf`? It doesn't verify the scheme, but should work for RSync objects.

karlkfi replied:
IDOf only works if the GK is set on the object, which isn't true of all objects. That convention is unique to Config Sync code and the k8sobjects libs, but it's not shared by upstream k8s code or client-go.

I found a similar lookup function in my fake client code, so I'm going to move it into the core package next to IDof, so it's easy to discover.

At some point in the future, we should require everything to check the scheme unless the object is Unstructured, but that would be a big refactor to handle the possible errors.

karlkfi replied:
Had to put it in kinds.LookupID to avoid a package dependency cycle

karlkfi commented on e2e/testcases/policy_dir_test.go:
```diff
@@ -50,30 +50,24 @@ func TestPolicyDirUnset(t *testing.T) {
-		nt.T.Fatal(err)
-	}
+	nomostest.SetRootSyncGitDir(nt, rootSyncID.Name, "")
+	nomostest.SetExpectedSyncPath(nt, rootSyncID, ".") // TODO: OCI default sync path?
```
TIL `controllers.DefaultSyncDir` and `gitproviders.DefaultSyncDir` are different, confusingly so.

nan-yu replied:
`controllers.DefaultSyncDir` is used by the reconciler in prod, while `gitproviders.DefaultSyncDir` is only used for testing.

karlkfi replied:
Replaced most of these with two in the OCI RSync constructors, tho those only work (for now) if the source was Git or OCI before. I'll add OCISourceSync construction in a followup PR.

This on specifically is an exception, since it's changed after construction, but I changed it to use the constant.

karlkfi commented on e2e/testcases/private_cert_secret_test.go:
```diff
@@ -421,20 +421,16 @@ func TestOCICACertSecretRefRootRepo(t *testing.T) {
 	nt.MustMergePatch(rs, fmt.Sprintf(`{"spec": {"sourceType": "%s", "oci": {"image": "%s", "auth": "none"}, "git": null}}`,
 		configsync.OciSource, imageURL))
-	nt.WaitForRootSyncSourceError(configsync.RootSyncName, status.SourceErrorCode, "tls: failed to verify certificate: x509: certificate signed by unknown authority")
+	nomostest.SetExpectedSyncPath(nt, rootSyncID, ".") // TODO: OCI default sync path?
```
IDOf only works if the GK is set on the object, which isn't true of all objects. That convention is unique to Config Sync code and the k8sobjects libs, but it's not shared by upstream k8s code or client-go.

I found a similar lookup function in my fake client code, so I'm going to move it into the core package next to IDof, so it's easy to discover.

At some point in the future, we should require everything to check the scheme unless the object is Unstructured, but that would be a big refactor to handle the possible errors.

nan-yu reviewed (approved):
/lgtm

nan-yu commented:
/retest

karlkfi commented:
last fixes:
- Updated TestOCICACertSecretRefNamespaceRepo to store the Git source and restore it after switching to OCI and back
- Updated TestHelmCACertSecretRefNamespaceRepo to store the Git source and restore it after switching to Helm and back
- Fixed TestMultiSyncs_Unstructured_MixedControl by fixing ResetExpectedGitSync to work with RepoSyncs

karlkfi commented:
I had planned to save this for another commit, but I had to add a SetExpectedOCIImage to fix TestOCICACertSecretRefNamespaceRepo, which was flipping between Git and OCI and back. There's still some more work to do to use OCISyncSources everywhere we sync from OCI and replace the use of WithRootSha1Func, but I'll save that for a followup PR.

nan-yu commented on e2e/nomostest/config_sync.go:
```diff
@@ -947,17 +931,66 @@ func RepoSyncObjectV1Beta1(nn types.NamespacedName, repoURL string, sourceFormat
+	}
+	SetExpectedOCISource(nt, id, imageID, expectedImageDigest, expectedSyncPath)
+	rs := k8sobjects.RootSyncObjectV1Beta1(name)
+	rs.Spec.SourceFormat = configsync.SourceFormatUnstructured // TODO: is SourceFormat necessary with OCI?
```
Technically, OCI source configs can use the hierarchical format. We don't have test coverage for that scenario. Given that unstructured is the recommended format, so we can skip testing OCI with hierarchical format.

karlkfi replied:
Replaced TODO with a comment in https://github.com/GoogleContainerTools/kpt-config-sync/pull/1410

nan-yu commented on e2e/nomostest/config_sync.go:
```diff
@@ -974,20 +1007,34 @@ func (nt *NT) RootSyncObjectOCI(name string, image *registryproviders.OCIImage)
-	rs := RepoSyncObjectV1Beta1FromNonRootRepo(nt, nn)
+	SetExpectedOCISource(nt, id, imageID, expectedImageDigest, expectedSyncPath)
+	rs := k8sobjects.RepoSyncObjectV1Beta1(nn.Namespace, nn.Name)
+	rs.Spec.SourceFormat = configsync.SourceFormatUnstructured // TODO: is SourceFormat necessary with OCI?
```
RepoSync always uses unstructured format.

karlkfi replied:
We don't need to specify it for RepoSyncs, right?

nan-yu replied:
Correct. The default value is unstructured for RepoSyncs.

nan-yu reviewed (approved):
/lgtm
